tests/ui/redundant_type_annotations.rs

```diff
@@ -0,0 +1,67 @@
+#![allow(unused)]
+#![warn(clippy::redundant_type_annotations)]
+
+struct A;
+enum E {
+    One,
+    Two,
+}
+
+fn f() -> String {
+    String::new()
+}
+
+fn f_struct() -> A {
+    A
+}
+
+fn f_enum() -> E {
+    E::One
+}
+
+struct ATest2 {
+    inner: u32,
+}
+
+impl ATest2 {
+    fn func(&self) -> u32 {
+        self.inner
+    }
+
+    fn a(&self) {
+        let v: u32 = self.func(); // This should lint but doesn't (MethodCall)
+    }
+
+    fn get_num() -> u32 {
+        5
+    }
+
+    fn new() -> Self {
+        Self { inner: 5 }
+    }
+}
+
+fn f_prim() -> u32 {
+    5
+}
+
+fn main() {
```

Comment:
I would suggest a few more tests. It's fine if not all of them work right away, but I definitely want to prevent false positives (FPs)

```rs
#[derive(Debug, Default)]
struct Cake<T> {
    _data: T,
}

fn make_something<T: Default>() -> T {
    T::default()
}

fn make_cake<T: Default>() -> Cake<T> {
    Cake::<T>::default()
}

fn test_generics() {
    // The type annotation is needed to determine T
    let _c: Cake<i32> = make_something();

    // The type annotation is needed to determine the topic
    let _c: Cake<u8> = make_cake();
}

fn test_non_locals() {
    // This shouldn't lint
    fn _arg(x: u32) -> u32 {
        x
    }

    // This could lint, but probably shouldn't
    let _closure_arg = |x: u32| x;
}

fn test_complex_types() {
    // Shouldn't lint, since the literal will be i32 otherwise
    let _u8: u8 = 128;

    // Should lint
    let _tuple_i32: (i32, i32) = (12, 13);

    // Shouldn't lint, since the tuple will be i32 otherwise
    let _tuple_u32: (u32, u32) = (1, 2);

    // Should lint, since the type is determined by the init value
    let _tuple_u32: (u32, u32) = (3_u32, 4_u32);

    // Should lint
    let _array: [i32; 3] = [5, 6, 7];

    // Shouldn't lint
    let _array: [u32; 2] = [8, 9];
}
```

I also feel like it would be better, to split up the main method into functions, which specify if and why an example should be linted :)